src/scenic/core/regions.py

```diff
@@ -281,883 +306,2816 @@ def __hash__(self):
-		minx, maxx = findMinMax(x)
-		miny, maxy = findMinMax(y)
-		for x in range(minx, maxx+1):
-			for y in range(miny, maxy+1):
-				p = self.gridToPoint((x, y))
-				if self.grid[y, x] == 1 and obj.containsPoint(p):
-					return False
-		return True
+    """A Region given by an obstacle grid.
+
+    A point is considered to be in a `GridRegion` if the nearest grid point is
+    not an obstacle.
+
+    Args:
+        name (str): name for debugging
+        grid: 2D list, tuple, or NumPy array of 0s and 1s, where 1 indicates an obstacle
+          and 0 indicates free space
+        Ax (float): spacing between grid points along X axis
+        Ay (float): spacing between grid points along Y axis
+        Bx (float): X coordinate of leftmost grid column
+        By (float): Y coordinate of lowest grid row
+        orientation (`VectorField`; optional): orientation of region
+    """
+    def __init__(self, name, grid, Ax, Ay, Bx, By, orientation=None):
+        self.grid = numpy.array(grid)
+        self.sizeY, self.sizeX = self.grid.shape
+        self.Ax, self.Ay = Ax, Ay
+        self.Bx, self.By = Bx, By
+        y, x = numpy.where(self.grid == 0)
+        points = [self.gridToPoint(point) for point in zip(x, y)]
+        super().__init__(name, points, orientation=orientation)
+
+    def gridToPoint(self, gp):
+        x, y = gp
+        return ((self.Ax * x) + self.Bx, (self.Ay * y) + self.By)
+
+    def pointToGrid(self, point):
+        x, y, z = point
+        x = (x - self.Bx) / self.Ax
+        y = (y - self.By) / self.Ay
+        nx = int(round(x))
+        if nx < 0 or nx >= self.sizeX:
+            return None
+        ny = int(round(y))
+        if ny < 0 or ny >= self.sizeY:
+            return None
+        return (nx, ny)
+
+    def containsPoint(self, point):
+        gp = self.pointToGrid(point)
+        if gp is None:
+            return False
+        x, y = gp
+        return (self.grid[y, x] == 0)
+
+    def containsObject(self, obj):
+        # TODO improve this procedure!
+        # Fast check
+        for c in obj.corners:
+            if not self.containsPoint(c):
+                return False
+        # Slow check
+        gps = [self.pointToGrid(corner) for corner in obj.corners]
+        x, y = zip(*gps)
+        minx, maxx = findMinMax(x)
+        miny, maxy = findMinMax(y)
+        for x in range(minx, maxx+1):
+            for y in range(miny, maxy+1):
+                p = self.gridToPoint((x, y))
+                if self.grid[y, x] == 1 and obj.containsPoint(p):
+                    return False
+        return True
+
+    def containsRegionInner(self, reg, tolerance):
+        raise NotImplementedError
+
+###################################################################################################
+# View Regions
+###################################################################################################
+
+class DefaultViewRegion(MeshVolumeRegion):
+    """ The default view region shape.
```

Comment:
I don't think this is particularly informative. How about calling it `FieldOfViewRegion` (or maybe just `ViewRegion`) and describing it as "The viewing volume of a camera defined by a radius and horizontal/vertical view angles."? It's true this region type is used as the default visible region for objects, but that's not how it's defined -- it's better to describe it intrinsically.